src/librustc_typeck/check/method/mod.rs

```diff
@@ -369,21 +370,45 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                         span: Span,
                         method_name: ast::Ident,
                         self_ty: Ty<'tcx>,
+                        self_ty_hir: &hir::Ty,
                         expr_id: ast::NodeId)
                         -> Result<Def, MethodError<'tcx>> {
+        debug!("resolve_ufcs: method_name={:?} self_ty={:?} expr_id={:?}",
+            method_name,
+            self_ty,
+            expr_id
+        );
+
+        let tcx = self.tcx;
+
+        // Check if we have an enum variant here.
+        if let ty::Adt(adt_def, _) = self_ty.sty {
+            if adt_def.is_enum() {
+                if allow_type_alias_enum_variants(tcx, self_ty_hir, span) {
+                    let variant_def = adt_def.variants.iter().find(|vd| {
+                        tcx.hygienic_eq(method_name, vd.ident, adt_def.did)
+                    });
+                    if let Some(variant_def) = variant_def {
+                        let def = Def::VariantCtor(variant_def.did, variant_def.ctor_kind);
+                        return Ok(def);
```

Comment:
Hmm, I think stability check is missing (`check_stability`).
And perhaps privacy check as well, not sure.

Reply:
All other associated items are privacy and stability checked somewhere during name resolution in `librustc_typeck`, but not variants.

Reply:
Right. What do I want here exactly? Something like this?

```
tcx.check_stability(def.def_id(), Some(ref_id), span);
```

I'm pretty sure the def-id is right, but not sure about the node-id or span (do we just want the ones passed toe the fn?).

How do I check privacy?

I presume I don't need to stability/privacy-check the enum itself, only the variant.

Reply:
>Something like this?

Yep.
There's an example for privacy as well below (`if !item.vis.is_accessible_from(def_scope, tcx) { ... }`).

>I presume I don't need to stability/privacy-check the enum itself, only the variant.

Yes, for enums it's already checked elsewhere.

Reply:
Argh, the example below was about `associated_path_def_to_ty`.

Reply:
Privacy checking for associated values is actually much deeper inside `fn probe_for_name` and private candidates are actually skipped during search, rather than immediately create errors.

But that's all unnecessary right now, just check'n'error would be good enough for now.
This is all relevant only for code using macros 2.0 anyway, without them you can't refer to a private variant without referring to a private enum first.

Reply:
Enum variants aren't ever checked for privacy right now, it seems. So, it looks like I don't need to bother with this. You may want to file a separate issue about this however, especially insofar as it relates to macros.